arrow-flight/examples/flight_sql_server.rs

```diff
@@ -617,7 +618,7 @@ impl FlightSqlService for FlightSqlServiceImpl {
         &self,
         _query: CommandPreparedStatementQuery,
         _request: Request<PeekableFlightDataStream>,
-    ) -> Result<Response<<Self as FlightService>::DoPutStream>, Status> {
+    ) -> Result<Option<DoPutPreparedStatementResult>, Status> {
```

Comment:
Isn't it the field on `DoPutPreparedStatementResult` that is optional? It seems like this API would be cleaner and follow the rest of the API if it returned a DoPutPreparedStatementResult (and the client can handle / translate a missing message into a DoPutPreparedStatementResult)?

```suggestion
    ) -> Result<DoPutPreparedStatementResult, Status> {
```

Or maybe it is important to distinguish between the case where the server didn't send back any message from the case where the server sent back a message with an empty prepared handle?

Reply:
Yes I think I had some confusion around whether or not the response should now be mandatory in up-to-date servers, with the legacy behavior being something only the client is concerned with. 

But now I think making the new result explicitly optional helps to document the legacy behavior and helps avoid confusion. So I will update interfaces to reflect this.

Reply:
But yes to answer your question, the result is potentially missing due to legacy behavior and the types here indicate that directly. This is because the inner handle was originally specified to be *mandatory* if the result is provided. I've updated the protobuf in the [spec PR](apache/arrow#40243) to make the handle field itself `optional`.

So with that protobuf change, what we could do is make the inner handle field an `Option` and use a `Default` implementation on `DoPutPreparedStatementResult` to handle the case where it's missing,

Reply:
This is a decision of whether or not we want to expressly forbid up-to-date servers from implementing the older version of the spec, even though clients still support it. I went with the more conservative decision of allowing existing implementations to continue doing what they were doing before without any major changes aside from changing their return value to `Ok(None)`. That way they continue with the legacy behavior without being "locked out" of future crate updates.